File: 2025-09/2025-09-20/file_storage.py
def number_of_files(file_size, file_unit, drive_size_gb):
    
    return int(drive_size_gb / (
        file_size / 1000 if file_unit == "MB" 
        else file_size / 1000000 if file_unit == "KB" 
        else file_size / 1000000000))

print(number_of_files(500, "KB", 1))
print(number_of_files(50000, "B", 1))
print(number_of_files(5, "MB", 1))
print(number_of_files(4096, "B", 1.5))
print(number_of_files(220.5, "KB", 100))
print(number_of_files(4.5, "MB", 750))

# --- TEST SUITE ---

tests_text = '''
1. number_of_files(500, "KB", 1) should return 2000.
2. number_of_files(50000, "B", 1) should return 20000.
3. number_of_files(5, "MB", 1) should return 200.
4. number_of_files(4096, "B", 1.5) should return 366210.
5. number_of_files(220.5, "KB", 100) should return 453514.
6. number_of_files(4.5, "MB", 750) should return 166666.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 09 September/09-20/fileStorage.js
function numberOfFiles(fileSize, fileUnit, driveSizeGb) {
  
  return Math.floor(driveSizeGb / (
    fileUnit === "MB" ? fileSize / 1000
    : fileUnit === "KB" ? fileSize / 1000000
    : fileSize / 1000000000));
}

console.log(numberOfFiles(500, "KB", 1));
console.log(numberOfFiles(50000, "B", 1));
console.log(numberOfFiles(5, "MB", 1));
console.log(numberOfFiles(4096, "B", 1.5));
console.log(numberOfFiles(220.5, "KB", 100));
console.log(numberOfFiles(4.5, "MB", 750));

// --- TEST SUITE ---

const testsText = `
1. numberOfFiles(500, "KB", 1) should return 2000.
2. numberOfFiles(50000, "B", 1) should return 20000.
3. numberOfFiles(5, "MB", 1) should return 200.
4. numberOfFiles(4096, "B", 1.5) should return 366210.
5. numberOfFiles(220.5, "KB", 100) should return 453514.
6. numberOfFiles(4.5, "MB", 750) should return 166666.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);